### PR TITLE
[INFRA/CORE] Implement Actions + some tidying up

### PR DESCRIPTION
**Pull request by lhl2617**
# Summary

- Implement actions as per spec and discussion with @yp717
- Every org (IIFO, ...) has a list of resulting actions
- At EoT (end of turn) islands can submit any extra actions
- These actions are then dispatched by `GameState.DispatchActions` to mutate the state.
- Note that `action` strictly refers to actions that can be initiated by islands/via orgs.

## Additional Information
- Wrap errors properly to give more context
- Remove unused module in `go.mod` (using `go mod tidy`) -- module seems to be added by @Neelesh99 but usage was removed

## Test Plan

Added a sample action (still in this code change) check logs for `
2020/12/09 15:20:42 [GAMESTATE]: Team1 gave 1 resources to Team2`.

- NOTICE THAT THIS IS REMOVED PRE-MERGE!

- `go test ./...`
```
?       github.com/SOMAS2020/SOMAS2020  [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team1   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team2   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team3   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team4   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team5   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team6   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common  (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/config   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common/rules    (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/shared   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/server  (cached)
?       github.com/SOMAS2020/SOMAS2020/pkg/testutils    [no test files]
```

- `go run .`
```2020/12/09 15:20:42 [SERVER]: start runTurn
2020/12/09 15:20:42 [SERVER]: TURN: 1, Season: 1
2020/12/09 15:20:42 [SERVER]: start updateIsland
2020/12/09 15:20:42 [Team3]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:20:42 [Team4]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:20:42 [Team5]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:20:42 [Team6]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:20:42 [Team1]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:20:42 [Team2]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/09 15:20:42 [SERVER]: finish updateIsland
2020/12/09 15:20:42 [SERVER]: start runOrgs
2020/12/09 15:20:42 [SERVER]: start runIITO
2020/12/09 15:20:42 [SERVER]: finish runIITO
2020/12/09 15:20:42 [SERVER]: start runIIFO
2020/12/09 15:20:42 [SERVER]: finish runIIFO
2020/12/09 15:20:42 [SERVER]: start runIITO
2020/12/09 15:20:42 [SERVER]: finish runIITO
2020/12/09 15:20:42 [SERVER]: finish runOrgs
2020/12/09 15:20:42 [SERVER]: start getEndOfTurnActions
2020/12/09 15:20:42 [Team3]: EndOfTurnActions
2020/12/09 15:20:42 [Team4]: EndOfTurnActions
2020/12/09 15:20:42 [Team5]: EndOfTurnActions
2020/12/09 15:20:42 [Team6]: EndOfTurnActions
2020/12/09 15:20:42 [Team1]: EndOfTurnActions
2020/12/09 15:20:42 [Team2]: EndOfTurnActions
2020/12/09 15:20:42 [SERVER]: finish getEndOfTurnActions
2020/12/09 15:20:42 [GAMESTATE]: start DispatchActions
2020/12/09 15:20:42 [GAMESTATE]: Team1 gave 1 resources to Team2
2020/12/09 15:20:42 [GAMESTATE]: finish DispatchActions
2020/12/09 15:20:42 [SERVER]: start endOfTurn
2020/12/09 15:20:42 [SERVER]: start probeDisaster
2020/12/09 15:20:42 [SERVER]: finish probeDisaster
2020/12/09 15:20:42 [SERVER]: start deductCostOfLiving
2020/12/09 15:20:42 [SERVER]: finish deductCostOfLiving
2020/12/09 15:20:42 [SERVER]: start updateIslandLivingStatus
2020/12/09 15:20:42 [SERVER]: finish updateIslandLivingStatus
2020/12/09 15:20:42 [SERVER]: finish endOfTurn
2020/12/09 15:20:42 [SERVER]: finish runTurn
2020/12/09 15:20:42 [SERVER]: start runTurn
2020/12/09 15:20:42 [SERVER]: TURN: 2, Season: 1
2020/12/09 15:20:42 [SERVER]: start updateIsland
2020/12/09 15:20:42 [Team2]: Received game state update: {1 2 map[Team1:{89 Alive 0} Team2:{91 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:20:42 [Team3]: Received game state update: {1 2 map[Team1:{89 Alive 0} Team2:{91 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:20:42 [Team4]: Received game state update: {1 2 map[Team1:{89 Alive 0} Team2:{91 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:20:42 [Team5]: Received game state update: {1 2 map[Team1:{89 Alive 0} Team2:{91 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:20:42 [Team6]: Received game state update: {1 2 map[Team1:{89 Alive 0} Team2:{91 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:20:42 [Team1]: Received game state update: {1 2 map[Team1:{89 Alive 0} Team2:{91 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/09 15:20:42 [SERVER]: finish updateIsland
2020/12/09 15:20:42 [SERVER]: start runOrgs
2020/12/09 15:20:42 [SERVER]: start runIITO
2020/12/09 15:20:42 [SERVER]: finish runIITO
2020/12/09 15:20:42 [SERVER]: start runIIFO
2020/12/09 15:20:42 [SERVER]: finish runIIFO
2020/12/09 15:20:42 [SERVER]: start runIITO
2020/12/09 15:20:42 [SERVER]: finish runIITO
2020/12/09 15:20:42 [SERVER]: finish runOrgs
2020/12/09 15:20:42 [SERVER]: start getEndOfTurnActions
2020/12/09 15:20:42 [Team2]: EndOfTurnActions
2020/12/09 15:20:42 [Team3]: EndOfTurnActions
2020/12/09 15:20:42 [Team4]: EndOfTurnActions
2020/12/09 15:20:42 [Team5]: EndOfTurnActions
2020/12/09 15:20:42 [Team6]: EndOfTurnActions
2020/12/09 15:20:42 [Team1]: EndOfTurnActions
2020/12/09 15:20:42 [SERVER]: finish getEndOfTurnActions
2020/12/09 15:20:42 [GAMESTATE]: start DispatchActions
2020/12/09 15:20:42 [GAMESTATE]: Team1 gave 1 resources to Team2
2020/12/09 15:20:42 [GAMESTATE]: finish DispatchActions
2020/12/09 15:20:42 [SERVER]: start endOfTurn
2020/12/09 15:20:42 [SERVER]: start probeDisaster
2020/12/09 15:20:42 [SERVER]: finish probeDisaster
2020/12/09 15:20:42 [SERVER]: start deductCostOfLiving
2020/12/09 15:20:42 [SERVER]: finish deductCostOfLiving
2020/12/09 15:20:42 [SERVER]: start updateIslandLivingStatus
2020/12/09 15:20:42 [SERVER]: finish updateIslandLivingStatus
2020/12/09 15:20:42 [SERVER]: finish endOfTurn
2020/12/09 15:20:42 [SERVER]: finish runTurn
2020/12/09 15:20:42 [SERVER]: Max turns '2' reached or exceeded
===== START OF TURN 1 (END OF TURN 0) =====
common.GameState{Season:0x1, Turn:0x1, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 2 (END OF TURN 1) =====
common.GameState{Season:0x1, Turn:0x2, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:89, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:91, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 3 (END OF TURN 2) =====
common.GameState{Season:0x1, Turn:0x3, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:78, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:82, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
```

### CODE REVIEW

**Review by yp717 — COMMENTED** [review]
This makes sense as a base for the actions and I think the get and dispatch action methods can probably support most of what is in the mvp/draft

File: internal/server/turn.go [hunk 1]
```diff
@@ -17,7 +18,37 @@ func (s *SOMASServer) runTurn() error {
...
+	}
+
+	// get all end of turn actions
+	if actions, err := s.getEndOfTurnActions(); err != nil {
+		return err
+	} else {
+		allActions = append(allActions, actions...)
+	}
+
+	// !!!!!!!!!!!!
```
Comment (lhl2617) [inline internal/server/turn.go:37]: Leaving this here for now--to be removed before merging.

File: internal/common/action.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+	"github.com/pkg/errors"
+)
+
+// ActionType is an enum that classifies actions by their type
+type ActionType int
+
+const (
+	// GiveClientResource : give resource from a client to the other
```
Comment (lhl2617) [inline internal/common/action.go:12]: this is _strictly_ an example to show how it's done

File: internal/server/turn.go [hunk 3]
```diff
@@ -14,37 +15,76 @@ func (s *SOMASServer) runTurn() error {
...
+		},
+	}
+	allActions = append(allActions, exampleAction)
+
+	// dispatch actions
+	if err := s.gameState.DispatchActions(allActions); err != nil {
+		return errors.Errorf("Error dispatching actions: %v", err)
 	}
 
 	if err := s.endOfTurn(); err != nil {
```
Comment (lhl2617) [inline internal/server/turn.go:55]: yes, in here we mutate `gameState` using the server but not a function of `gameState`, but this is strictly for convenience only--actions are strictly defined as items initiated by islands + the org meetings. Server is free to mutate gameState as it wishes as long as it is a valid mutation.

**Review by QFSW — APPROVED** [review]

**Pull request merged**